tracer/src/Datadog.Trace/Tagging/TraceTagCollection.cs

```diff
@@ -0,0 +1,189 @@
+// <copyright file="TraceTagCollection.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using Datadog.Trace.Util;
+
+namespace Datadog.Trace.Tagging
+{
+    internal class TraceTagCollection
+    {
+        // key1=value1,key2=value2
+        private const char TagPairSeparator = ',';
+        private const char KeyValueSeparator = '=';
+
+        // tags with this prefix are propagated horizontally
+        // (i.e. from upstream services and to downstream services)
+        // using the `x-datadog-tags` header
+        private const string PropagatedTagPrefix = "_dd.p.";
+
+        // for now we only expect one trace-level tag:
+        // "_dd.p.upstream_services"
+        private const int DefaultCapacity = 1;
+
+        // "_dd.p.a=b"
+        public const int MinimumPropagationHeaderLength = 9;
+        public const int MaximumPropagationHeaderLength = 512;
+
+        private static readonly char[] TagPairSeparators = { TagPairSeparator };
+
+        private List<KeyValuePair<string, string>> _tags;
+        private string? _cachedPropagationHeader;
+
+        public TraceTagCollection()
+        {
+            _tags = new List<KeyValuePair<string, string>>(DefaultCapacity);
+        }
+
+        public TraceTagCollection(List<KeyValuePair<string, string>> tags)
+        {
+            _tags = tags;
+        }
+
+        public static TraceTagCollection ParseFromPropagationHeader(string? propagationHeader)
+        {
+            if (string.IsNullOrEmpty(propagationHeader))
+            {
+                return new TraceTagCollection();
+            }
+
+            var tags = propagationHeader!.Split(TagPairSeparators, StringSplitOptions.RemoveEmptyEntries);
+            var tagList = new List<KeyValuePair<string, string>>(tags.Length);
+
+            foreach (var tag in tags)
+            {
+                var separatorIndex = tag.IndexOf(KeyValueSeparator);
+
+                // there must be at least one char before and
+                // one char after the separator ("a=b")
+                if (separatorIndex > 0 && separatorIndex < tag.Length - 2)
```

Comment:
I think the condition is slightly off. Can you also add this `"a=b"` example to the unit tests to be sure?

```suggestion
                if (separatorIndex > 0 && separatorIndex < tag.Length - 1)
```

Reply:
Good catch!

Reply:
Fixed and added test in 942aeb2cc9d8f95bdd155d7fd721acd7a9f2f616.